### PR TITLE
Allow selecting docker as cri.name

### PR DESCRIPTION
**Pull request by voelzmo**
**How to categorize this PR?**

/kind enhancement

**What this PR does / why we need it**:
Allow selecting `docker` as `cri.name`, such that users can explicitly choose `docker` once we switched the default to `containerd`. This decouples the choice of the container runtime from the next k8s update and allows for switching back to `docker` without downgrading your k8s version.

**Which issue(s) this PR fixes**:
related #4110

**Release note**:

```feature operator
Allow explicit configuration of `docker` as a container runtime (`.spec.provider.workers[].cri.name` field in `Shoot`s) for backwards compatibility. Select this only if your workload doesn't run nicely with `containerd`. This configuration option will be removed in the future!
```

/cc @BeckerMax

### CODE REVIEW

**Review by rfranzke — CHANGES_REQUESTED** [review]
Due to https://github.com/gardener/gardener/blob/77e09dd522d7bfb682cc35d0be437201b606061b/extensions/pkg/controller/worker/machines.go#L146-L152, I think this will trigger a(n undesired) rolling update of the shoot worker nodes if somebody with `.spec.provider.workers[].cri=nil` starts to specify `.spec.provider.workers[].cri.name=docker` (although, it's semantically equal).

I guess we need some special handling in above mentioned function, right?

**Comment by rfranzke** [comment]
cc @grolu @petersutter @holgerkoser

**Comment by voelzmo** [comment]
> Due to
> 
> https://github.com/gardener/gardener/blob/77e09dd522d7bfb682cc35d0be437201b606061b/extensions/pkg/controller/worker/machines.go#L146-L152
> 
> , I think this will trigger a(n undesired) rolling update of the shoot worker nodes if somebody with `.spec.provider.workers[].cri=nil` starts to specify `.spec.provider.workers[].cri.name=docker` (although, it's semantically equal).
> I guess we need some special handling in above mentioned function, right?

That's a great observation, I should have put this context in the above summary! @BeckerMax and I decided to do this in a separate change, because technically it is an optimization and can be treated as a separate concern. I've added this to the tracking issue (#4110) and we'll get to it later, does this make sense?

**Comment by voelzmo** [comment]
A side note: In order to be able to select `docker` as `cri.name`, you'd have to edit your `CloudProfile` such that the used `MachineImage` contains `docker` in the list of suppported CRIs. This seems actually useful, because we want to remove `docker` from the images at a future point in time, actually resulting in images which are no longer compatible with the choice of `docker` as container runtime.

The error message you get could be more helpful to figure out what to do, though
```
shoots.core.gardener.cloud "shoot-gcp-contd" is forbidden: [spec.provider.workers[0].cri.name: Unsupported value: "docker": supported values: "containerd"]
```

**Review by rfranzke — DISMISSED** [review]
OK, sounds reasonable, thank you!

/lgtm

**Comment by voelzmo** [comment]
Discussing with @timebertt we came to the conclusion this still needs an additional change. 
Do not merge yet!

File: pkg/apis/core/validation/shoot.go
Comment (ialidzhikov) [inline pkg/apis/core/validation/shoot.go:72]: With such change, I think `docs/extensions/operatingsystemconfig.md` needs an update.

There are expressions like
> The only CRI supported at the moment is: "containerd"

> If the `.spec.cri` section exists then the `name` property is mandatory. The only valid value at the moment is `containerd`.

that now need an update.

Comment (ialidzhikov) [inline pkg/apis/core/validation/shoot.go:72]: You can also extend `example/30-cloudprofile.yaml` and `example/90-shoot.yaml` and show that it is possible to set `docker` as cri name.

Comment (voelzmo) [reply pkg/apis/core/validation/shoot.go:72]: > With such change, I think docs/extensions/operatingsystemconfig.md needs an update.

Ah, this is where we hide this configuration, thanks for the pointer! I was looking at the API docs to find out more about supported values for `CRI.name`, but there wasn't anything mentioned. This is most likely out of the scope for this change, but do you feel `operatingsystemconfig.md` is the right place to mention the supported options for `CRI.name`?

Comment (voelzmo) [reply pkg/apis/core/validation/shoot.go:72]: I feel comfortable adding this to `example/30-cloudprofile.yaml` to list the possible configuration options and also hinting that people add `docker` to their `MachineImageVersion`s. For `example/90-shoot.yaml` I'm not quite sure if we'd want to do this:
* This would either mean adding a second `Worker` to the example or introduce some "either/or" configuration for the worker we currently have in there. Neither feels right from the point of view of keeping an example "simple".
* We probably don't want to actively encourage people to choose this configuration. 
WDYT?

Comment (ialidzhikov) [reply pkg/apis/core/validation/shoot.go:72]: > I was looking at the API docs to find out more about supported values for CRI.name, but there wasn't anything mentioned.

We can mention the allowed values in the doc string of the corresponding field. For me, such change makes sense.

Comment (ialidzhikov) [reply pkg/apis/core/validation/shoot.go:72]: > We probably don't want to actively encourage people to choose this configuration.
WDYT?

Ah, ok, agree. Let's then don't mention about `docker` in `example/90-shoot.yaml`.

Comment (ialidzhikov) [reply pkg/apis/core/validation/shoot.go:72]: > We can mention the allowed values in the doc string of the corresponding field. For me, such change makes sense.

Do you want to do it as part of this PR or do you want to leave it to another PR?

Comment (voelzmo) [reply pkg/apis/core/validation/shoot.go:72]: I think a separate change would be better, given that #4222 also changes API docs and includes a `make generate` run.

**Review by rfranzke — APPROVED** [review]
/lgtm
/reviewed ok-to-test
/assign @ialidzhikov

**Review by vpnachev — COMMENTED** [review]
Overall looks good.
I think you have to add one improvement in the worker pool hash https://github.com/gardener/gardener/blob/77e09dd522d7bfb682cc35d0be437201b606061b/extensions/pkg/controller/worker/machines.go#L148-L150

```golang
if w.CRI != nil && w.CRI.Name != "docker" {
	data = append(data, string(w.CRI.Name))
}
```

This way the nodes rolling update will be avoided when `worker.cri=nil` is re-configured as `worker.cri.name=docker`.

**Comment by voelzmo** [comment]
> Overall looks good.
> I think you have to add one improvement in the worker pool hash
> [...]
> This way the nodes rolling update will be avoided when `worker.cri=nil` is re-configured as `worker.cri.name=docker`.

Thanks for spotting this! As [discussed with @rfranzke above](https://github.com/gardener/gardener/pull/4218#issuecomment-862321984) we're doing this in an extra change: #4237

**Review by vpnachev — APPROVED** [review]
Thanks! I haven't seen #4237 on time.
/lgtm

**Pull request merged**
